---
title: "Processing Stool Sample Whole Genome Sequencing Data"
author: "Ankan Choudhury"
date: "2025-02-19"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{bash}

```


```{r packages}

library(biomformat)
library(tidyverse)
library(data.table)
library(fs)
library(phyloseq)
library(writexl)
library(phylostratr)
library(ape)
library(taxize)
```

```{r Kaiju data transformation}

Kaijubioms<-dir("Kaiju Taxonomy Abundance",pattern="*biom")
Kaijutaxa<-dir("Kaiju Taxonomy Abundance",pattern="kaiju_taxonomy*")
for (biom in Kaijubioms){
  for(taxa in Kaijutaxa){
    if(substring(biom,21,nchar(biom)-5)==substring(taxa,16,nchar(taxa)-4))
      name=substring(biom,21,nchar(biom)-5)
    OTUtable<-as.data.frame(as.matrix(biom_data(read_biom(paste("Kaiju Taxonomy Abundance",biom,sep="/")))))%>%
      rownames_to_column("Feature.ID")
    taxatable<-read.table(paste("Kaiju Taxonomy Abundance",taxa,sep="/"),sep='\t',header = TRUE)
    Abundance<-inner_join(OTUtable,taxatable)%>%
      select(-"Feature.ID")%>%
      relocate(Taxon)%>%
      mutate(Taxon=str_replace_all(Taxon,".__",""))%>%
      separate(Taxon,c("Domain","Phylum","Class","Order","Family","Genus","Species"),sep=";")%>%
      mutate(Phylum=if_else(is.na(Phylum),Domain,Phylum),
             Class=if_else(is.na(Class),Phylum,Class),
             Order=if_else(is.na(Order),Class,Order),
             Family=if_else(is.na(Family),Order,Family),
             Genus=if_else(is.na(Genus),Family,Genus),
             Species=if_else(is.na(Species),Genus,Species))
    write_xlsx(Abundance,paste("Kaiju Taxonomy Abundance/kaiju_abun-",name,".xlsx",sep=""))
  }
}

head(select(as.data.frame(OTUtable),c(1:10)))
head(taxatable)
head(select(as.data.frame(Abundance),c(1:10)))
```

```{r Creating Kaiju phyloseq object}

Kaijubioms<-dir("Kaiju Taxonomy Abundance",pattern="*biom")
Kaijutaxa<-dir("Kaiju Taxonomy Abundance",pattern="kaiju_taxonomy*")
dir.create("PhyloseqOutputs/KaijuPhyloseq")
for (biom in Kaijubioms){
  for(taxa in Kaijutaxa){
    if(substring(biom,21,nchar(biom)-5)==substring(taxa,16,nchar(taxa)-4))
      name=substring(biom,21,nchar(biom)-5)
    OTUtable<-as.matrix(biom_data(read_biom(paste("Kaiju Taxonomy Abundance",biom,sep="/"))))
    taxatable<-read.table(paste("Kaiju Taxonomy Abundance",taxa,sep="/"),sep='\t',header = TRUE)%>%
      mutate(Taxon=str_replace_all(Taxon,".__",""))%>%
      separate(Taxon,c("Domain","Phylum","Class","Order","Family","Genus","Species"),sep=";")%>%
      mutate(Phylum=if_else(is.na(Phylum),Domain,Phylum),
             Class=if_else(is.na(Class),Phylum,Class),
             Order=if_else(is.na(Order),Class,Order),
             Family=if_else(is.na(Family),Order,Family),
             Genus=if_else(is.na(Genus),Family,Genus),
             Species=if_else(is.na(Species),Genus,Species))%>%
      column_to_rownames("Feature.ID")%>%
      as.matrix()
    physeq <- phyloseq(otu_table(OTUtable, taxa_are_rows = TRUE), tax_table(taxatable))
    saveRDS(physeq,paste("PhyloseqOutputs/KaijuPhyloseq/phyloseq_kaiju-",name,".rds",sep=""))
  }
}

summary(physeq)
physeq



```

```{r Creating Bracken Taxonomic abundance Table}

brackentables<-dir(path="BrackenOutputs/Do",pattern="table-bracken*")
brackentaxonomy<-dir(path="BrackenOutputs/Do",pattern="taxonomy-bracken*")

dir.create("Kraken-Bracken Taxonomy Abundance")
for(each in brackentables){
  name<-substring(each,15)
  for(each2 in brackentaxonomy){
    name2<-substring(each2,18)
    if(name==name2)
    {table<-as.data.frame(as.matrix(biom_data(read_biom(paste("BrackenOutputs/Do",each,"feature-table.biom",sep='/')))))%>%
      rownames_to_column("Feature.ID")
    taxonomy<-read.table(paste("BrackenOutputs/Do",each2,"taxonomy.tsv",sep='/'),sep="\t",header = TRUE)%>%
      mutate(Feature.ID=as.character(Feature.ID))
    taxatable<-inner_join(table,taxonomy)%>%
      select(-"Feature.ID")%>%
      relocate(Taxon)%>%
      mutate(Taxon=str_replace_all(Taxon,".__",""))%>%
      separate(Taxon,c("Domain","Kingdom","Phylum","Class","Order","Family","Genus","Species"),sep=";")
    write_xlsx(taxatable,paste("Kraken-Bracken Taxonomy Abundance/Kraken-Bracken_abun-",name,".xlsx",sep=""))
    }
  }
}
head(select(as.data.frame(table),c(1:10)))
head(as.data.frame(taxonomy))
head(select(as.data.frame(taxatable),c(1:10)))
```


```{r Creating Bracken Phyloseq}

brackentables<-dir(path="BrackenOutputs/Do",pattern="table-bracken*")
brackentaxonomy<-dir(path="BrackenOutputs/Do",pattern="taxonomy-bracken*")
brackentree<-dir(path="BrackenOutputs/Do",pattern="bracken-reports*")

dir.create("PhyloseqOutputs/BrackenPhyloseq")
counter=0
for(each in brackentables){
  name<-substring(each,15)
  table<-as.matrix(biom_data(read_biom(paste("BrackenOutputs/Do",each,"feature-table.biom",sep='/'))))
  taxonomy<-read.table(paste("BrackenOutputs/Do/taxonomy-bracken-",name,"/taxonomy.tsv",sep=''),sep="\t",header = TRUE)%>%
    mutate(Taxon=str_replace_all(Taxon,".__",""))%>%
    separate(Taxon,c("Domain","Phylum","Class","Order","Family","Genus","Species"),sep=";")%>%
    mutate(Phylum=if_else(is.na(Phylum),Domain,Phylum),
           Class=if_else(is.na(Class),Phylum,Class),
           Order=if_else(is.na(Order),Class,Order),
           Family=if_else(is.na(Family),Order,Family),
           Genus=if_else(is.na(Genus),Family,Genus),
           Species=if_else(is.na(Species),Genus,Species))%>%
    column_to_rownames("Feature.ID")%>%
    as.matrix()
  tree<-read_tree(paste("BrackenOutputs/Do/bracken-reports-",name,"/output.tree",sep=''))
  counter=counter+1
  print(counter)
  physeqBracken<-phyloseq(otu_table(table,taxa_are_rows = TRUE),tax_table(taxonomy))
  physeqBracken<-merge_phyloseq(physeqBracken,tree)
  saveRDS(physeqBracken,paste("PhyloseqOutputs/BrackenPhyloseq/phyloseq_bracken-",name,".rds",sep=""))
  print(name)
}

head(select(as.data.frame(table),c(1:10)))
head(as.data.frame(taxonomy))
summary(physeqBracken)
physeqBracken


```
```{r Creating Metaphlan Phyloseq}

MetaphlanOutputs<-dir("MetaphlanOutput")
dir.create("PhyloseqOutputs/MetaphlanPhyloseq")
MetaphlanTree<-read_tree("20201013_mpav30_speciesMod.nwk")
new_names <- MetaphlanTree$tip.label
new_names<-new_names%>%
  str_replace_all("s__","")
MetaphlanTree$tip.label<-new_names
for(folder in MetaphlanOutputs){
  Alltaxonomy<-NULL
  AllOTU<-NULL
  for(file in list.files(paste("MetaphlanOutput",folder,sep="/"))){
    content<-readLines(paste("MetaphlanOutput",folder,file,sep="/"))[-c(1:4)]
    id<-substring(basename(file),1,nchar(basename(file))-11)
    totalreads<-readLines(paste("MetaphlanOutput",folder,file,sep="/"))[3]
    totalreads<-as.numeric(gsub("^#(\\d+).*", "\\1", totalreads))
    table<-read.table(text=content,header=FALSE,sep="\t")%>%
      select(-4)%>%
      mutate(V1=str_replace_all(V1,".__",""),
             V1 = str_replace_all(V1, "\\|", ";"),
             reads=round(V3/100*totalreads)) %>%
      separate(V1,c("Kingdom","Phylum","Class","Order","Family","Genus","Species"),sep=";")%>%
      separate(V2,c("KID","PID","CID","OID","FID","GID","SID"),sep="\\|")%>%
      unique()%>%
      filter(!(is.na(Species)))%>%
      mutate(PID=if_else(PID==""|is.na(PID),KID,PID),
             CID=if_else(CID==""|is.na(CID),PID,CID),
             OID=if_else(OID==""|is.na(OID),CID,OID),
             FID=if_else(FID==""|is.na(FID),OID,FID),
             GID=if_else(GID==""|is.na(GID),FID,GID),
             SID=if_else(SID==""|is.na(SID),GID,SID))
    taxonomy<-table[,c(1:7)]
    OTU<-table[,c(7,16)]%>%
      unique()%>%
      mutate(Sample=id)
    Alltaxonomy<-rbind(Alltaxonomy,taxonomy)
    AllOTU<-rbind(AllOTU,OTU)
    print(paste("Processed: MetaphlanOutput",folder,file,sep="/"))
  }
  
  taxa_table<-Alltaxonomy%>%
    arrange("Species")%>%
    unique()
  rownames(taxa_table)<-taxa_table$Species
  otu_table<-AllOTU%>%
    group_by(Sample,Species)%>%
    summarize(reads=sum(reads))%>%
    pivot_wider(id_cols = "Species",values_from = "reads",names_from = "Sample",values_fill = 0)%>%
    column_to_rownames("Species")
  physeqMetaphlan<-phyloseq(otu_table(as.matrix(otu_table),taxa_are_rows = TRUE),tax_table(as.matrix(taxa_table)))
  physeqMetaphlan<-merge_phyloseq(physeqMetaphlan,phy_tree(MetaphlanTree))
  saveRDS(physeqMetaphlan,paste("PhyloseqOutputs/MetaphlanPhyloseq/phyloseq_Metaphlan-",substring(folder,1,(nchar(folder)-7)),".rds",sep=""))
}

head(taxa_table)
head(select(as.data.frame(otu_table),c(1:10)))
```
```{r Creating Humann Phyloseq}

HumannOutput<-dir("HumannOutput")
dir.create("PhyloseqOutputs/HumannPhyloseq")
for (folder in HumannOutput){
  AllCoverage<-data.frame()
  AllAbundance<-data.frame()
  counter=1
  for(file in list.files(paste("HumannOutput",folder,sep="/"),pattern="*_pathabundance.tsv")){
    id<-substring(basename(file),1,nchar(basename(file))-18)
    abundance<-read.delim(paste("HumannOutput",folder,file,sep="/"),sep='\t',header = TRUE,comment.char = "")
    df<-colnames(abundance)%>%
      str_replace_all("X\\.\\.","")%>%
      str_replace_all("X","")%>%
      str_replace_all("\\.","_")%>%
      str_replace_all("_Abundance","")
    colnames(abundance)<-df
    coverage<-read.delim(paste("HumannOutput/",folder,"/",id,"_pathcoverage.tsv",sep=""),sep='\t',header = TRUE,comment.char = "")
    df<-colnames(coverage)%>%
  str_replace_all("X\\.\\.","")%>%
  str_replace_all("X","")%>%
  str_replace_all("\\.","-")%>%
  str_replace_all("_Coverage","")
    colnames(coverage)<-df
    if (counter==1){
      AllAbundance<-abundance}
    else{AllAbundance<-left_join(AllAbundance,abundance)}
    if(counter==1){
    AllCoverage<-coverage}
    else{AllCoverage<-left_join(AllCoverage,coverage)}
    counter=counter+1
      print(paste("Processed",folder,":",id))
  }
  
  AllAbundance<-AllAbundance%>%
    column_to_rownames("Pathway")%>%
    as.matrix()
  AllCoverage<-AllCoverage%>%
    column_to_rownames("Pathway")%>%
    as.matrix()
  physeqcoverage<-phyloseq(otu_table(AllCoverage,taxa_are_rows=TRUE))
  physeqabundance<-phyloseq(otu_table(AllAbundance,taxa_are_rows=TRUE))
  saveRDS(physeqabundance,paste("PhyloseqOutputs/HumannPhyloseq/phyloseq_humann-",folder,"_pathwayabundance.rds",sep=""))
  saveRDS(physeqcoverage,paste("PhyloseqOutputs/HumannPhyloseq/phyloseq_humann-",folder,"_pathwaycoverage.rds",sep=""))
}

for (folder in HumannOutput){
  AllRPK<-data.frame()
  counter=1
  for(file in list.files(paste("HumannOutput",folder,sep="/"),pattern="*_genefamilies.tsv")){
    id<-substring(basename(file),1,nchar(basename(file))-17)
    RPK<-read.delim(paste("HumannOutput",folder,file,sep="/"),sep='\t',header = TRUE,comment.char = "")
    df<-colnames(RPK)%>%
      str_replace_all("X\\.\\.","")%>%
      str_replace_all("X","")%>%
      str_replace_all("\\.","-")%>%
      str_replace_all("_Abundance-RPKs","")
    colnames(RPK)<-df
    if (counter==1){
      AllRPK<-RPK}
    else{AllRPK<-left_join(AllRPK,RPK)}
    counter=counter+1
    print(paste("Processed",folder,":",id))
  }
  
  AllRPK<-AllRPK%>%
    column_to_rownames("Gene-Family")%>%
    as.matrix()
  physeqRPK<-phyloseq(otu_table(AllRPK,taxa_are_rows=TRUE))
  saveRDS(physeqRPK,paste("PhyloseqOutputs/HumannPhyloseq/phyloseq_humann-",folder,"_genefamilies.rds",sep=""))
  
}

head(abundance)
head(select(as.data.frame(AllAbundance),c(1:10)))
head(coverage)
head(select(as.data.frame(AllCoverage),c(1:10)))
head(RPK)
head(select(as.data.frame(AllRPK),c(1:10)))
rm(AllRPK)
```
## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
